[DTensor] compute shape and offset for arbitrary _StridedShard

weifengpy opened the pull request:
resolve https://github.com/pytorch/pytorch/issues/167859

for _StridedShard, compute_local_shape_and_global_offset was landed to consider fsdp2 + tp: (_StridedShard(0, split_factor=mesh.size(k)), Shard(0)). Need to extend it to arbitrary _StridedShard

for example, `_StridedShard(dim=0, split_factor=batch_size), _StridedShard(dim=0, split_factor=batch_size * seq_len / device_mesh.size(0))`

This PR ensure correct local shape for DTensor views with _StridedShard

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #166483
* __->__ #168146

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D87897203](https://our.internmc.facebook.com/intern/diff/D87897203)

weifengpy committed:
support arbitrary _StridedShard
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "support arbitrary _StridedShard"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "support arbitrary _StridedShard"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "support arbitrary _StridedShard"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "support arbitrary _StridedShard"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "support arbitrary _StridedShard"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "support arbitrary _StridedShard"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "support arbitrary _StridedShard"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "support arbitrary _StridedShard"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy commented on torch/distributed/tensor/_utils.py:
```diff
@@ -178,83 +127,60 @@ def _compute_local_shape_and_global_offset(
-            ):
-                zero_global_offset = torch.SymInt(zero_global_offset)
-
-            global_offset[shard_dim] = torch.sym_ite(
```
@ezyang,  do you know if torch.SymInt are still needed? which unit test triggers torch.SymInt? I didn't include torch.SymInt / torch.sym_ite logic in the PR. CI is doing well. Seems to be related to LocalTensor

weifengpy replied:
nvm. I can trigger torch.SymInt from test/distributed/tensor/test_dtensor.py. It's indeed for LocalTensor. will fix it

wconstab commented on torch/distributed/tensor/_utils.py:
```diff
@@ -178,83 +127,60 @@ def _compute_local_shape_and_global_offset(
+        if shard_size == 0:
+            shard_dim_to_global_offsets[shard_dim] = [zero_global_offset]
+            continue
+        if not isinstance(placement, _StridedShard):
```
i was asking about the actual behavior not the isinstance check.  but now that you mention the isinstance, i think its worth asking ourselves "what's going to happen if someone adds "MyNewSharding".  I think your logic needs to be specific to SS and S, and might not be possible to make work for arbitrary new shardings?

wconstab replied:
Well, can you check? We should have DCp test coverage but I'd like you to make sure we aren't breaking DCP

ezyang commented on torch/distributed/tensor/_utils.py:
```diff
@@ -161,6 +109,7 @@ def _compute_local_shape_and_global_offset(
     mesh_shape: ShapeType,
     my_coordinate: Optional[list[int]],
     placements: Sequence[Placement],
+    skip_offset: bool = False,
```
Can you please document this

weifengpy replied:
sounds good. will add docstring for skip_offset

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859

for _StridedShard, compute_local_shape_and_global_offset was landed to consider fsdp2 + tp: (_StridedShard(0, split_factor=mesh.size(k)), Shard(0)). Need to extend it to arbitrary _StridedShard

for example, `_StridedShard(dim=0, split_factor=batch_size), _StridedShard(dim=0, split_factor=batch_size * seq_len / device_mesh.size(0))`

This PR ensure correct local shape for DTensor views with _StridedShard


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859

for _StridedShard, compute_local_shape_and_global_offset was landed to consider fsdp2 + tp: (_StridedShard(0, split_factor=mesh.size(k)), Shard(0)). Need to extend it to arbitrary _StridedShard

for example, `_StridedShard(dim=0, split_factor=batch_size), _StridedShard(dim=0, split_factor=batch_size * seq_len / device_mesh.size(0))`

This PR ensure correct local shape for DTensor views with _StridedShard


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859

for _StridedShard, compute_local_shape_and_global_offset was landed to consider fsdp2 + tp: (_StridedShard(0, split_factor=mesh.size(k)), Shard(0)). Need to extend it to arbitrary _StridedShard

for example, `_StridedShard(dim=0, split_factor=batch_size), _StridedShard(dim=0, split_factor=batch_size * seq_len / device_mesh.size(0))`

This PR ensure correct local shape for DTensor views with _StridedShard


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge -f "doc build failure is unrelated"

yangw-dev commented:
@pytorchbot revert -m "failed internal tests  due to AttributeError: 'LocalIntNode' object has no attribute 'int_', please fix it and re-merge again" -c ghfirst

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859

for _StridedShard, compute_local_shape_and_global_offset was landed to consider fsdp2 + tp: (_StridedShard(0, split_factor=mesh.size(k)), Shard(0)). Need to extend it to arbitrary _StridedShard

for example, `_StridedShard(dim=0, split_factor=batch_size), _StridedShard(dim=0, split_factor=batch_size * seq_len / device_mesh.size(0))`

This PR ensure correct local shape for DTensor views with _StridedShard


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy commented:
> @pytorchbot revert -m "failed internal tests due to AttributeError: 'LocalIntNode' object has no attribute 'int_', please fix it and re-merge again" -c ghfirst

re-merging as I verified the error is gone with the latest diff

weifengpy committed:
Update on "[DTensor] compute shape and offset for arbitrary _StridedShard"

resolve https://github.com/pytorch/pytorch/issues/167859

for _StridedShard, compute_local_shape_and_global_offset was landed to consider fsdp2 + tp: (_StridedShard(0, split_factor=mesh.size(k)), Shard(0)). Need to extend it to arbitrary _StridedShard

for example, `_StridedShard(dim=0, split_factor=batch_size), _StridedShard(dim=0, split_factor=batch_size * seq_len / device_mesh.size(0))`

This PR ensure correct local shape for DTensor views with _StridedShard


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D87897203](https://our.internmc.facebook.com/intern/diff/D87897203)

[ghstack-poisoned]

tianyu-l commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -684,12 +684,49 @@ def _to_replicate_tensor(
+
+    @staticmethod
+    @maybe_run_for_local_tensor
+    def local_shard_size_and_offset(  # pyre-ignore[bad-override]
```
What's the purpose of having this function? It's not normal that a private method `_local_shard_size_and_offset` calls into a public method.

tianyu-l commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -684,12 +684,49 @@ def _to_replicate_tensor(
+    @staticmethod
+    @maybe_run_for_local_tensor
+    def local_shard_size_and_offset(  # pyre-ignore[bad-override]
+        self,
```
why a staticmethod would have `self` as the first arg?

tianyu-l reviewed (commented):
Sorry for the late review. Please see if my comments make sense to you.
